Do not allow manual changes to manageFields via subresources

nodo opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
It prevents users to manually change manageFields via a subresource. If a request tries to do that we return the managedField of the live object.

**Which issue(s) this PR fixes**:
Fixes #89274

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

nodo commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -66,44 +66,45 @@ type Manager interface {
 // on update and apply requests.
-func NewFieldManager(f Manager) *FieldManager {
-	return &FieldManager{f}
+func NewFieldManager(f Manager, i bool) *FieldManager {
```
Sure thing!

nodo commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -122,16 +123,22 @@ func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (o
-			managed = internal.NewEmptyManaged()
+	} else {
+		managed, err = internal.DecodeObjectManagedFields(newAccessor.GetManagedFields())
+		if err != nil || len(managed.Fields()) == 0 || f.isSubresource {
```
That's a good point! I will think about something following your example.

nodo commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -139,15 +140,15 @@ func (f *TestFieldManager) Reset() {
 
 func (f *TestFieldManager) Apply(obj runtime.Object, manager string, force bool) error {
-	out, err := fieldmanager.NewFieldManager(f.fieldManager).Apply(f.liveObj, obj, manager, force)
+	out, err := fieldmanager.NewFieldManager(f.fieldManager, f.isSubresource).Apply(f.liveObj, obj, manager, force)
```
I am totally fine to have a new struct `TestSubresourceFieldManager` which embeds a `TestFieldManager`. However the method above become something like:

```go
func (f *TestFieldManager) Apply(obj runtime.Object, manager string, force bool) error {
    out, err := fieldmanager.NewFieldManager(f.fieldManager, false).Apply(f.liveObj, obj, manager, force)
    // ....
}

I am not sure the hardcoded `false` feels better. I am not sure if I missed your point though.

nodo replied:
@apelisse which PR are we waiting for?

apelisse replied:
Sorry, that's kind of unfortunate. @julianvmodesto knows, he has multiple I'm not sure which it is. @nodo you could then probably rebase your change on top of his so that you can start making progress.

julianvmodesto replied:
Ah this one: https://github.com/kubernetes/kubernetes/pull/84134

I think it is close to merging.

See the field_manager_test.go, where we use an entire FieldManager in the test now.

fedebongio commented:
/assign @apelisse 
/cc @kwiesmueller

nodo committed:
Do not allow manual changes to manageFields via subresources

If a request tries to change managedFields, the response returns the
managedField of the live object.

nodo commented:
@apelisse Thanks for the review! I have rebased and addressed (to some extent) your comments:

- I have created a `NewSubresourceTestFieldManager` function to simplify a bit the unit-test, however `NewTestFieldManager` is still used in other places and that still takes an extra parameter.
- I have refactor a bit the logic to decode new and fallback on existing when necessary.

Let me know what you think when you have time.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -119,36 +120,54 @@ func newDefaultFieldManager(f Manager, typeConverter internal.TypeConverter, obj
-// use-case), and simply updates the managed fields in the output
-// object.
-func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (object runtime.Object, err error) {
+func decodeManagedFields(liveObj, newObj runtime.Object, isSubresource bool) (Managed, error) {
```
This function is only used once and doesn't really add any value, does it?

apelisse replied:
Yeah, I suspect we could do simpler than that somehow

nodo replied:
I see your point, I kind of like to have in a separate function because in case of an error we can just `return` and the logic is much less nested. Note that with this change we catch an error that was not handled before (see [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go#L149) ).

I have tried to follow your [previous suggestion](https://github.com/kubernetes/kubernetes/pull/91993#discussion_r438301338) but after `isResetManagedFields` has been added I didn't really like how the logic was flowing.

WDYT?

apelisse replied:
I suspect the bug you fixed is the intended behavior and you broke it? Can you describe what you think you've solved? I hope the tests will catch something.

I don't necessarily disagree that the separate function helps, but the content of this function can still be improved.

nodo replied:
oops you are totally right! I will try to simplify it a bit. I find the `if managed, err = ...` a bit confusing 😬

nodo replied:
> I hope the tests will catch something.
I think they didn't 😬

lavalamp replied:
> This function is only used once and doesn't really add any value, does it?

I think it's complicated enough that the function name is valuable documentation.

nodo commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -1577,7 +1577,110 @@ func TestErrorsDontFailPatch(t *testing.T) {
+	_, client, closeFn := setup(t)
+	defer closeFn()
+
+	podName := "just-a-pod"
```
yep, you are right, I have removed them.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -1093,3 +1097,52 @@ func getLastApplied(obj runtime.Object) (string, error) {
 }
+
+func TestUpdateViaSubresources(t *testing.T) {
+	f := NewSubresourceTestFieldManager(schema.FromAPIVersionAndKind("v1", "Pod"))
```
I think you can add the managedFields in the object just below when you run the first update, which would eliminate the need to run 2 "Update" (by the way, update doesn't take a patch, it takes the full patched object)

nodo replied:
aha! makes sense, I will change it! 👍

nodo replied:
Done in b4786b8

apelisse commented:
/ok-to-test

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -119,36 +120,59 @@ func newDefaultFieldManager(f Manager, typeConverter internal.TypeConverter, obj
 
+	// We take the managedFields of the live object in case the request tries to
+	// manually set managedFields via a subresource.
+	if isSubresource {
```
this should come before the "isResetManagedFields" method above. Otherwise, someone would be able to reset the managed fields through the subresource, which we probably don't want.

nodo replied:
Totally, thanks for catching it!

nodo replied:
Done in c8b4842, I have also added a unit test to make sure this case is covered.

lavalamp replied:
This check can be done first in this function, no? The accessor is not used if this is for a subresource?

nodo replied:
Agree, I think the subresource check should go first. Thanks for spotting it!

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -119,36 +120,59 @@ func newDefaultFieldManager(f Manager, typeConverter internal.TypeConverter, obj
-// use-case), and simply updates the managed fields in the output
-// object.
-func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (object runtime.Object, err error) {
+func decodeLiveManagedFields(liveObj runtime.Object) (Managed, error) {
```
I'm not a big fan, but I could move forward with that and then we can try to play with it and see what we like best :-)

nodo replied:
No worries, I am very happy to iterate on it :-) I am not super satisfied either

nodo commented:
/retest

apelisse commented:
Tests seem spurious, I'd address the "resetfieldmanager" being before the sub-resource check before trying to test any further.

nodo commented:
Originally, I was thinking to add an integration test for the CRD path on the same lines as [this one](https://github.com/kubernetes/kubernetes/pull/91993/files#diff-4ac5831d494b1b52c7c7be81e552a458R1582). What do you think?

apelisse commented:
> Originally, I was thinking to add an integration test for the CRD path on the same lines as this one. What do you think?

Yeah, that sounds reasonable.

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -1093,3 +1097,56 @@ func getLastApplied(obj runtime.Object) (string, error) {
+	expectedManager := "fieldmanager_test_subresource"
+	if err := f.Update(obj, expectedManager); err != nil {
+		t.Fatalf("failed to apply object: %v", err)
+	}
```
Why are you able to set the manager here, but not the second time? I'm confused.

apelisse replied:
OK No, I misread that's actually correct, thanks!

apelisse reviewed (commented):
I think that's much better, thanks! One sort-of nit?

apelisse commented:
/lgtm
/approve

nodo commented:
Thanks @apelisse. I have added an integration test for CRDs and squashed the commits. 

I am not sure why `triage/needs-information` got added though :)

nodo commented:
/assign @sttts

apelisse commented:
/unassign @sttts 
/assign @lavalamp

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -839,6 +839,7 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 
 			MaxRequestBodyBytes: r.maxRequestBodyBytes,
 		}
+		isSubresource := subresources != nil
```
I don't know this particular code-- does the main resource go through this code path? I would expect the main resource and subresources to come up with different values for this.

nodo replied:
>  does the main resource go through this code path?
Yes, from what I understand this is the case.

nodo replied:
Interesting, I need to investigate this path a bit more. Thanks for raising this.

apelisse replied:
Basically CRDs have the main resource which uses a `FieldManager` [1], the status sub-resource that uses the same FieldManager [3], and the scale subresource that uses no manager [4]. We clearly can't use the same FieldManager for main resource and status sub-resource so we'll have to create a new one that has the "ignore" flag explicitly set to true.

[1]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L844-L852
[2]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L878
[3]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L873

nodo replied:
I have fixed the problem by adding a separate `FieldManager` for the status subresource (https://github.com/kubernetes/kubernetes/pull/91993/commits/02c1948f71de628628034caf8c6d98bd1816ec5c). Given `scale` doesn't have a `FieldManager` yet I have left it as it is.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -119,36 +120,59 @@ func newDefaultFieldManager(f Manager, typeConverter internal.TypeConverter, obj
+		return decodeLiveManagedFields(liveObj)
+	}
+
+	if isResetManagedFields(newAccessor.GetManagedFields()) {
```
This can also go before constructing the accessor.

(It's just a little confusing to construct the accessor before we're sure it's going to be needed.)

nodo replied:
I am not sure I understood, I think `newAccessor` should be constructed at this point because `isResetManagedFields` needs to check the managed fields of the new object.

lavalamp commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -129,6 +130,50 @@ spec:
 	}
 	verifyReplicas(t, result, 1)
+
+	// Try to set managed fields using a subresource and verify that it has no effect
```
Is there a CRD test that tries to set the managed fields via the main resource? I'd like to see that that still works.

nodo replied:
What do you mean by "main resource"?

nodo replied:
Ah I see, you mean without the subresource path. 🤔 yeah, I can't see a test that explicitly tries and set the managed fields using CRDs.

nodo replied:
Added a new test here: https://github.com/kubernetes/kubernetes/pull/91993/commits/d9a577f73949bdd75bd8d0b2c5c1bd3fe192d1c6

lavalamp reviewed (commented):
I think this approach is OK for now.

I think the theoretically correct thing to do is permit this or not based on whether the subresource schema is exactly the same as the main resource schema (as in the case of status). If the Kind of the request doesn't match the main resource Kind, then the fieldsets don't make any sense at all, they won't match the object.

For that reason, I think we should consider omitting the field sets in data sent to the user when serving subresources.

If we ever come up with a use case for setting these via status, we can modify this to make the check about the Kind and not just isSubresource. (I don't foresee this, though.)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -67,18 +67,19 @@ type Manager interface {
 type FieldManager struct {
-	fieldManager Manager
+	fieldManager  Manager
+	isSubresource bool
```
I guess, considering what I just said, "ignoreRequestObjectManagedFields" is more accurately what this changes about the logic. That name is a mouthful, maybe you can think of a better one. Personally, I prefer my fields to be named for the effect they accomplish rather than the source of the decision. That ensures that somewhere a reader will see a line like `ignoreManagedFieldsFromRequestObject = isSubresource` and it will be very clear both what we are doing and why we are doing it.

nodo replied:
Thanks for the tip, sounds a good idea to me - I will make the change 👍

nodo commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -129,6 +130,50 @@ spec:
+	if err != nil {
+		t.Fatalf("Error updating subresource: %v ", err)
+	}
+	newManagedFields, err := getManagedFields(result)
```
I have tried to use `Do().Get()` but I was getting this error:

```
    apply_crd_test.go:173: Error updating subresource: no kind "WishIHadChosenNoxu" is registered for version "mygroup.example.com/v1beta1" in scheme "k8s.io/client-go/kubernetes/scheme/register.go:70"
```

After a bit of tries I have switched back to `DoRaw()`. Do you know what I could be missing?

Another reason to keep `DoRaw()` is that I would need to add a similar logic to the helper function to extract the existing managed fields.

apelisse replied:
OK Forget it, there must be some registering of the type needed, that's not that important.

nodo commented:
@lavalamp @apelisse thanks for your reviews! I have added a couple of commits that address your comments. Let me know what you think when you have time.

apelisse commented:
OK, I think that's better. @kwiesmueller do you want to have a look?

kwiesmueller commented:
Sure, I'll take a look right tomorrow morning.

kwiesmueller commented:
If that's too long, you can merge ahead.

kwiesmueller reviewed (approved):
/lgtm
/approve

Looks good, thanks @nodo !

kwiesmueller commented:
/retest

nodo commented:
Thanks @kwiesmueller !

nodo commented:
/retest

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -876,6 +877,19 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 		// override status subresource values
 		// shallow copy
 		statusScope := *requestScopes[v.Name]
+		statusScope.FieldManager, err = fieldmanager.NewDefaultCRDFieldManager(
```
Contain in a `if utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply) {`?

nodo replied:
Thanks for spotting it!

lavalamp commented:
OK, sorry for delay, this LGTM except for the missing feature flag check.

nodo commented:
/retest

lavalamp commented:
/approve
/lgtm

lavalamp commented:
/milestone v1.20

apelisse commented:
Awesome @nodo! I'm marking this as done, thanks again for fixing this problem!

nodo commented:
Nice! Thanks for your help and your reviews @apelisse  👍
